### PR TITLE
Delays

### PR DESCRIPTION
**Pull request by MollieS**

### CODE REVIEW

File: src/main/java/ttt/players/DelayedPlayer.java [hunk 1]
```diff
@@ -7,32 +7,47 @@
...
-    public DelayedPlayer(Player player) {
+    public DelayedPlayer(RandomPlayer player) {
         this.player = player;
+        this.delay = 1000;
+    }
+
+    public DelayedPlayer(PerfectPlayer player) {
+        this.player = player;
+        this.delay = 0;
     }
```
Comment (MollieS) [inline src/main/java/ttt/players/DelayedPlayer.java:16]: Adds different constructors for the different types of computer players

File: src/main/java/ttt/players/DelayedPlayer.java [hunk 2]
```diff
@@ -7,32 +7,47 @@
 public class DelayedPlayer implements Player {
 
     private Player player;
+    private Integer delay;
 
-    public DelayedPlayer(Player player) {
+    public DelayedPlayer(RandomPlayer player) {
```
Comment (gemcfadyen) [inline src/main/java/ttt/players/DelayedPlayer.java]: Referring to different concrete types here starts to violate OCP. (If a new type of player came along, you'd need to add another constructor here).

Instead, could the delay be passed into the constructor, and the PlayerFactory decide whether a Delayed Player is needed, and if so, pass in the relevant timeout? This way all the Player type checking would be in one place.

File: src/main/java/ttt/players/DelayedPlayer.java [hunk 3]
```diff
@@ -7,32 +7,47 @@
...
+        this.delay = 0;
     }
 
     public Marks getMark() {
         return player.getMark();
     }
 
     public Integer getLocation(Board board) throws Exception {
-        long startTime = System.currentTimeMillis();
         Integer location = player.getLocation(board);
```
Comment (gemcfadyen) [inline src/main/java/ttt/players/DelayedPlayer.java]: can this be int rather than Integer?

**Comment by MollieS** [comment]
@gemcfadyen definitely an OCP violation!  I will move the checks to PlayerFactory.

Originally getLocation returned either the move or null for an invalid move, which is why it was an Integer, but now that I am raising an exception for an invalid move I think I can change it to int.

Thank you!

**Pull request merged**